### PR TITLE
Unifying ranges

### LINKED ISSUE #1547: Unifying ranges
We have currently
- `MemorySpan` (which is also a misnomer) which ought to be called `Range`
- `CallDataInfo`
- and we don't use a singular structure for either "return data" nor for "return at ..."

I propose to refactor this to have
- `Range`'s that are just an `offset` and a `size`
- `MemoryRange`'s that are a `Range`, a `long contextNumber` it and a `Bytes data`, maybe as an abstract class
- `CallDataRange` `ReturnDataRange`, `OutputDataRange` and `ReturnAtRange` that extend it

Or something like that.

### PR DESCRIPTION
**Pull request by OlivierBBB**

### CODE REVIEW

File: arithmetization/src/main/java/net/consensys/linea/zktracer/module/Util.java [hunk 1]
```diff
@@ -168,25 +168,45 @@ public static short getTxTypeAsInt(TransactionType txType) {
...
+
+    checkArgument(offset >= 0, "Offset must be non-negative");
+    checkArgument(size >= 0, "Size must be non-negative");
 
     final int dataSize = data.size();
-    Bytes output = Bytes.repeat((byte) 0x0, size);
-
-    if (dataSize >= positionStart) {
-      if (dataSize >= (positionStart + size)) {
-        output = data.slice(positionStart, size);
```
Comment (OlivierBBB) [inline arithmetization/src/main/java/net/consensys/linea/zktracer/module/Util.java:184]: The condition `dataSize >= positionStart` is problematic as `Bytes slice(int i, int length);` throws an exception if `i >= length`. Also some simple reformatting.

File: arithmetization/src/main/java/net/consensys/linea/zktracer/types/MemoryRange.java
Comment (OlivierBBB) [inline arithmetization/src/main/java/net/consensys/linea/zktracer/types/MemoryRange.java:1]: The main change in this PR. The goal is to replace all ad-hoc ranges with a `MemoryRange`.

File: arithmetization/src/main/java/net/consensys/linea/zktracer/types/Range.java
Comment (OlivierBBB) [inline arithmetization/src/main/java/net/consensys/linea/zktracer/types/Range.java:1]: ras: renaming from `MemorySpan -> Range`. Two reasons:
- a **range** is a minimum value and a maximum value while a **span** is the difference between the two
- I want to reserve the word `MemoryRange` to something that actually points into _someone's_ memory

File: arithmetization/src/main/java/net/consensys/linea/zktracer/types/MemoryRange.java [hunk 2]
```diff
@@ -12,53 +12,103 @@
...
+    return contextNumber;
   }
 
-  public boolean isUint64() {
-    return length().isUInt64();
+  public Bytes extract() {
+    return range.isEmpty()
+        ? Bytes.EMPTY
+        : slice(rawData, safeLongToInt(range.offset()), safeLongToInt(range.size()));
   }
```
Comment (OlivierBBB) [inline arithmetization/src/main/java/net/consensys/linea/zktracer/types/MemoryRange.java:109]: This method e.g. extracts the relevant data from the `rawData` e.g.
- `callData` from the `callerMemorySnapshot`
- `returnData` from the `callerMemorySnapshot`

File: arithmetization/src/main/java/net/consensys/linea/zktracer/types/Range.java
Comment (OlivierBBB) [reply arithmetization/src/main/java/net/consensys/linea/zktracer/types/Range.java:1]: image

File: arithmetization/src/main/java/net/consensys/linea/zktracer/module/Util.java [hunk 3]
```diff
@@ -168,25 +168,45 @@ public static short getTxTypeAsInt(TransactionType txType) {
...
+  }
+
+  /**
+   * This method expects a "small-ish" long value and returns the corresponding int value.
+   *
+   * @param value
+   * @return
+   * @throws ArithmeticException
+   */
+  public static int safeLongToInt(long value) throws ArithmeticException {
```
Comment (letypequividelespoubelles) [inline arithmetization/src/main/java/net/consensys/linea/zktracer/module/Util.java]: Move this fct in conversions.java

File: arithmetization/src/main/java/net/consensys/linea/zktracer/module/Util.java [hunk 4]
```diff
@@ -168,25 +168,45 @@ public static short getTxTypeAsInt(TransactionType txType) {
...
+   * Extracts {@code size} many bytes from {@code data} starting at {@code offset}. If {@code data}
+   * ``runs out'' it substitutes the missing bytes with 0's.
    *
    * @param size Bytes of
    * @param data, right-padded with 0's if needed, starting from
-   * @param positionStart
+   * @param offset
    */
-  public static Bytes slice(Bytes data, int positionStart, int size) {
+  public static Bytes slice(Bytes data, int offset, int size) {
```
Comment (letypequividelespoubelles) [inline arithmetization/src/main/java/net/consensys/linea/zktracer/module/Util.java]: I was planning to do it in my PR. I'll have to resolve conflicts now 😅

**Review by letypequividelespoubelles — APPROVED** [review]
LGTM

File: arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/Hub.java [hunk 5]
```diff
@@ -1033,7 +1033,7 @@ void traceOpcode(MessageFrame frame) {
         }
 
         final boolean returnFromDeployment =
-            (this.opCode() == RETURN || this.currentFrame().isDeployment());
```
Comment (OlivierBBB) [inline arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/Hub.java]: No idea what happened here 😆

**Pull request merged**